feat: creating new tutorial for creating a spring ai app (#4852)
JetBrains/kotlin-web-site: docs/topics/spring-ai-guide.md

store documents in a vector database, and answer questions using context from those documents.

You will use the following tools during this tutorial:

* [Spring Boot](https://spring.io/projects/spring-boot) as the base to configure and run the web application.
* [Spring AI](https://spring.io/projects/spring-ai) to interact with the LLM and perform context-based retrieval.
* [IntelliJ IDEA](https://www.jetbrains.com/idea/) to generate the project and implement the application logic.
* [Qdrant](https://qdrant.tech/) as the vector database for similarity search.
* [Docker](https://www.docker.com/) to run Qdrant locally.
* [OpenAI](https://platform.openai.com) as the LLM provider.

## Before you start

1. Download and install the latest version of [IntelliJ IDEA Ultimate Edition](https://www.jetbrains.com/idea/download/index.html).

    > If you use IntelliJ IDEA Community Edition or another IDE, you can generate a Spring Boot project using
    > a [web-based project generator](https://start.spring.io/#!language=kotlin&type=gradle-project-kotlin).
    >
    {style="tip"}

> 2. Create an OpenAI API key on the [OpenAI platform](https://platform.openai.com/api-keys) that you will use to access the API.

Reviewer: ```suggestion
2. Create an OpenAI API key on the [OpenAI platform](https://platform.openai.com/api-keys) to access the API.
```

3. Install [Docker](https://www.docker.com/) to run the Qdrant vector database locally.
4. After installing Docker, open your terminal and run the following command to start the container:

    ```bash
    docker run -p 6333:6333 -p 6334:6334 qdrant/qdrant
    ```

## Create the project

> > You can also use [Spring Boot's web-based project generator](https://start.spring.io/) as an alternative to generate your project.

Reviewer: ```suggestion
> You can use [Spring Boot web-based project generator](https://start.spring.io/) as an alternative to generate your project.
```

Reviewer: also + alternative

   > The IDE will generate and open a new project. It may take some time to download and import the project dependencies.
   >
   {style="tip"}

8. After this, you can see the following structure in the **Project view**:

   ![Spring Boot project view](spring-ai-project-view.png){width=400}

   The generated Gradle project corresponds to the Maven's standard directory layout:
    * There are packages and classes under the `main/kotlin` folder that belong to the application.
    * The entry point to the application is the `main()` method of the `SpringAiDemoApplication.kt` file.

## Update the project configuration

1. Update your `build.gradle.kts` Gradle build file with the following:

    ```kotlin
    plugins {
        kotlin("jvm") version "%kotlinVersion%"
        kotlin("plugin.spring") version "%kotlinVersion%"
        // Rest of the plugins
    }
   ```

2. Update your `springAiVersion` to `1.0.0-M6`:

   ```kotlin
   extra["springAiVersion"] = "1.0.0-M6"
   ```

3. Click the **Sync Gradle Changes** button to synchronize the Gradle files.
4. Update your `src/main/resources/application.properties` file with the following:

   ```text
   # OpenAI

>    spring.ai.openai.api-key=YOUR_OPENAI_API_KEY

Reviewer: Let's add a note after this code block with text like that: 
Set your OpenAI API key to the spring.ai.openai.api-key property 
or smth like that?

   spring.ai.openai.chat.options.model=gpt-4o-mini
   spring.ai.openai.embedding.options.model=text-embedding-ada-002
   # Qdrant
   spring.ai.vectorstore.qdrant.host=localhost
   spring.ai.vectorstore.qdrant.port=6334
   spring.ai.vectorstore.qdrant.collection-name=kotlinDocs
   spring.ai.vectorstore.qdrant.initialize-schema=true
   ```

5. Run the `SpringAiDemoApplication.kt` file to start the Spring Boot application.
   Once it's running, open the [Qdrant collections](http://localhost:6333/dashboard#/collections) page in your browser to see the following result:
   ![Qdrant collections](qdrant-collections.png){width=700}

## Create a controller to load and search documents

> Let's create a Spring `@RestController` to search documents and store them in the Qdrant collection:

Reviewer: ```suggestion
Create a Spring `@RestController` to search documents and store them in the Qdrant collection:
```

> 1. In the `src/main/kotlin/org/example/springaidemo` directory, create a new file named `KotlinSTDController.kt`, and add the following to it:

Reviewer: ```suggestion
1. In the `src/main/kotlin/org/example/springaidemo` directory, create a new file named `KotlinSTDController.kt`, and add the following code:
```

   ```kotlin
   package org.example.springaidemo
   
   import org.springframework.boot.autoconfigure.SpringBootApplication
   import org.springframework.boot.runApplication
   import org.springframework.context.annotation.Bean
   import org.springframework.web.client.RestTemplate
   
   @SpringBootApplication
   class SpringAiDemoApplication {
   
       @Bean
       fun restTemplate(): RestTemplate = RestTemplate()
   }
   
   fun main(args: Array<String>) {
       runApplication<SpringAiDemoApplication>(*args)
   }
   ```
   {collapsible="true"}

3. Run the application.
4. In the terminal, send a POST request to the `/kotlin/load-docs` endpoint to load the documents:

   ```bash
   curl -X POST http://localhost:8080/kotlin/load-docs
   ```

5. Once the documents are loaded, you can search for them with a GET request:

   ```Bash
   curl -X GET http://localhost:8080/kotlin/docs
   ```

   ![GET request results](spring-ai-get-results.png){width="700"}

> You can also view the results on the [Qdrant collections](http://localhost:6333/dashboard#/collections) page.
> 
{style="tip"}

## Implement an AI chat endpoint

> Once the documents are loaded, the final step is to add an endpoint that answers questions using the documents in Qdrant through Spring AI’s Retrieval-Augmented Generation (RAG) support:

Reviewer: ```suggestion
Once the documents are loaded, the final step is to add an endpoint that answers questions using the documents in Qdrant through Spring AI's Retrieval-Augmented Generation (RAG) support:
```

1. Open the `KotlinSTDController.kt` file, and import the following classes:

   ```kotlin
   import org.springframework.ai.chat.client.ChatClient
   import org.springframework.ai.chat.client.advisor.RetrievalAugmentationAdvisor
   import org.springframework.ai.chat.client.advisor.SimpleLoggerAdvisor
   import org.springframework.ai.chat.prompt.Prompt
   import org.springframework.ai.chat.prompt.PromptTemplate
   import org.springframework.ai.rag.preretrieval.query.transformation.RewriteQueryTransformer
   import org.springframework.ai.rag.retrieval.search.VectorStoreDocumentRetriever
   import org.springframework.web.bind.annotation.RequestBody
   ```

2. Add `ChatClient.Builder` to the controller's constructor parameters:

   ```kotlin
   class KotlinSTDController(
       private val chatClientBuilder: ChatClient.Builder,
       private val restTemplate: RestTemplate,
       private val vectorStore: VectorStore,
   )
   ```

> 3. Inside the controller class, create a `ChatClient` and a query transformer:

Reviewer: ChatClient instance?

                       .topK(request.topK)
                       .vectorStore(vectorStore)
                       .build()
               )
               .queryTransformers(rqtBuilder.promptTemplate(promptTemplate).build())
               .build()
   
           // Sends the prompt to the LLM with the retrieval advisor and get the response
           val response = chatClient.prompt(prompt)
               .advisors(retrievalAdvisor)
               .call()
               .content()
           logger.info("Chat response generated for query: '${request.query}'")
           return response
       }
   ```

5. Run the application.
6. In the terminal, send a POST request to the new endpoint to see the results:

   ```bash
   curl -X POST "http://localhost:8080/kotlin/chat/ask" \
        -H "Content-Type: application/json" \
        -d '{"query": "What are the performance implications of using lazy sequences in Kotlin for large datasets?", "topK": 3}'
   ```

   ![OpenAI answer to chat request](open-ai-chat-endpoint.png){width="700"}

Congratulations! You now have a Kotlin app that connects to OpenAI and answers questions using context retrieved from
documentation stored in Qdrant.

> Try experimenting with different queries or importing other documents to explore more possibilities.

Reviewer: Feel free to add links to the GitHub repo or to the projects mentioned there.
